<a href="https://colab.research.google.com/github/hansjang/Py_Radar/blob/main/%EB%B9%85%EB%B6%84%EA%B8%B0Part5_Exam6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!git clone https://github.com/AnalyticsKnight/yemoonsaBigdata

Cloning into 'yemoonsaBigdata'...
remote: Enumerating objects: 709, done.
remote: Counting objects: 100% (348/348), done.
remote: Compressing objects: 100% (233/233), done.
remote: Total 709 (delta 157), reused 237 (delta 115), pack-reused 361 (from 1)
Receiving objects: 100% (709/709), 11.46 MiB | 7.63 MiB/s, done.
Resolving deltas: 100% (341/341), done.


# 1. 작업형 제1유형

## 01. 시간 계산

In [4]:
import pandas as pd
import numpy as np
from tabulate import tabulate

df = pd.read_csv("/content/yemoonsaBigdata/datasets/Part5/ex6/06.01.01-Fire Station Data.csv", encoding="cp949")

print(tabulate(df.head(), headers="keys", tablefmt="grid"))
print(df.info())

+----+------------+------------------+------------------+------------------+
|    | 소방서ID   | 신고접수시간     | 출발시간         | 도착시간         |
+====+============+==================+==================+==================+
|  0 | StationS   | 2023-07-08 21:05 | 2023-07-08 21:09 | 2023-07-08 21:37 |
+----+------------+------------------+------------------+------------------+
|  1 | StationI   | 2023-06-04 8:11  | 2023-06-04 8:14  | 2023-06-04 8:58  |
+----+------------+------------------+------------------+------------------+
|  2 | StationJ   | 2023-01-31 16:20 | 2023-01-31 16:23 | 2023-01-31 16:44 |
+----+------------+------------------+------------------+------------------+
|  3 | StationD   | 2023-03-10 17:31 | 2023-03-10 17:32 | 2023-03-10 18:03 |
+----+------------+------------------+------------------+------------------+
|  4 | StationB   | 2023-06-18 7:23  | 2023-06-18 7:26  | 2023-06-18 8:18  |
+----+------------+------------------+------------------+------------------+
<class 'pandas.c

In [11]:
## 문자타입(object)을 날짜타입(datetime64)으로 변경하기
df1 = df.copy()
df1['신고접수시간'] = pd.to_datetime(df1['신고접수시간'])
df1['출발시간'] = pd.to_datetime(df1['출발시간'])
df1['도착시간'] = pd.to_datetime(df1['도착시간'])

## 출동시간 구하기
df1['출동시간'] = df1['도착시간'] - df1['출발시간']
## 출동시간을 초로 변환하기
df1['출동시간'] = df1['출동시간'].dt.total_seconds()

## 소방서별 출동시간 평균 구하기
df2 = df1.groupby('소방서ID')[['출동시간']].mean()
df2 = df2.reset_index()

## 출동시간 평균으로 내림차순 정렬
df2 = df2.sort_values('출동시간', ascending=False)

print(tabulate(df2.head(), headers="keys", tablefmt="grid"))

+----+------------+------------+
|    | 소방서ID   |   출동시간 |
+====+============+============+
| 25 | StationZ   |    2276.25 |
+----+------------+------------+
| 13 | StationN   |    2106.86 |
+----+------------+------------+
| 12 | StationM   |    2010    |
+----+------------+------------+
|  3 | StationD   |    2003.64 |
+----+------------+------------+
| 22 | StationW   |    1988.78 |
+----+------------+------------+



> **정답 : StationZ, 2276.25초**









## 02. 학생수와 교사수
### -> 교사 1인당 학생수가 가장 많은 학교와 그 학교의 전체학생 수  

In [12]:
import pandas as pd
import numpy as np
from tabulate import tabulate

df = pd.read_csv("/content/yemoonsaBigdata/datasets/Part5/ex6/06.01.02-Elementary School Data.csv")

print(tabulate(df.head(), headers="keys", tablefmt="grid"))
print(df.info())

+----+---------------+---------------+---------------+---------------+---------------+---------------+---------------+------------+
|    | School Name   |   1G Students |   2G Students |   3G Students |   4G Students |   5G Students |   6G Students |   Teachers |
+====+===============+===============+===============+===============+===============+===============+===============+============+
|  0 | ES1           |            64 |            48 |            84 |            78 |            24 |            38 |         15 |
+----+---------------+---------------+---------------+---------------+---------------+---------------+---------------+------------+
|  1 | ES2           |            58 |            62 |            82 |            73 |            42 |            63 |         15 |
+----+---------------+---------------+---------------+---------------+---------------+---------------+---------------+------------+
|  2 | ES3           |            64 |            79 |            99 |      

In [17]:
## 교사 1인당 학생수 구하기
df1 = df.copy()
df1['Students Tota1'] = df1['1G Students']+df1['2G Students']+df1['3G Students']+df1['4G Students']+df1['5G Students']+df1['6G Students']
df1['SPT'] = df1['Students Tota1']/df1['Teachers'] # SPT = Students Per Teachers


## SPT 기준으로 내림차순 정렬
df1 = df1.sort_values('SPT', ascending=False)

print(tabulate(df1.head(), headers="keys", tablefmt="grid"))

+-----+---------------+---------------+---------------+---------------+---------------+---------------+---------------+------------+------------------+---------+
|     | School Name   |   1G Students |   2G Students |   3G Students |   4G Students |   5G Students |   6G Students |   Teachers |   Students Tota1 |     SPT |
+=====+===============+===============+===============+===============+===============+===============+===============+============+==================+=========+
|  86 | ES87          |            38 |            66 |            21 |            24 |            61 |            35 |          9 |              245 | 27.2222 |
+-----+---------------+---------------+---------------+---------------+---------------+---------------+---------------+------------+------------------+---------+
|  57 | ES58          |            58 |            84 |            38 |            29 |            47 |            42 |         11 |              298 | 27.0909 |
+-----+---------------+-----

## 정답 : ES87, 245명

## 03. 경찰서별 범죄 데이터
### -> 연평균 범죄율(2022년/2021년)이 가장 많이 늘어난 경찰서의2021년 교통범죄(Traffic) 데이터

In [18]:
import pandas as pd
import numpy as np
from tabulate import tabulate

df = pd.read_csv("/content/yemoonsaBigdata/datasets/Part5/ex6/06.01.03-Crime Data.csv")

print(tabulate(df.head(), headers="keys", tablefmt="grid"))
print(df.info())

+----+------------------+--------------+-----------+---------+-----------+----------------+------------------+--------------------+--------+----------+-----------------+---------+------------+------------+------------+---------+
|    | Police Station   | Year-Month   |   Violent |   Theft |   Traffic |   Intelligence |   Against Morals |   Special Economic |   Drug |   Health |   Environmental |   Labor |   Security |   Election |   Military |   Other |
+====+==================+==============+===========+=========+===========+================+==================+====================+========+==========+=================+=========+============+============+============+=========+
|  0 | PS 1             | 2021-01-31   |        36 |      42 |        42 |             20 |               12 |                 27 |     35 |       26 |               7 |      16 |          6 |         18 |         13 |      17 |
+----+------------------+--------------+-----------+---------+-----------+----------

In [21]:
## 연도별 컬럼 추가
df1 = df.copy()
df1['Year-Month'] = pd.to_datetime(df1['Year-Month'])
df1['Year'] = df1['Year-Month'].dt.year

### 범죄 합계 추가하기
df1.loc[:, 'Total'] = df.loc[:, 'Violent':'Other'].sum(axis =1)

### 경찰서별, 연도별 트래픽, 토탈 범죄 데이터 합산
df2 = df1[['Police Station', 'Year', 'Traffic', 'Total']].groupby(by=['Police Station', 'Year']).sum(['Traffic', 'Total']).reset_index()

### 두 개 행을 열로 붙이기
df21 = df2[df2['Year'] == 2021][['Police Station', 'Year', 'Total']]
df22 = df2[df2['Year'] == 2022][['Police Station', 'Year', 'Total']]
df3 = pd.merge(df21, df22, on='Police Station')

# print(tabulate(df3.head(), headers="keys", tablefmt="grid"))

### 가장 많이 늘어난 경찰서 오름차순 정력
df3['Increase'] = df3['Total_y'] - df3['Total_x']
df3 = df3.sort_values('Increase', ascending = False)
### 가장 많이 늘어난 경찰서 가져오기
most_increase_station = df3[df3['Increase'] ==
										max(df3['Increase'])]['Police Station'].item()
### 가장 많이 늘어난 경찰서의 2021년 교통 범죄(Traffic) 수
result = df2[(df2['Police Station'] == most_increase_station) &
				(df2['Year'] == 2021)]['Traffic'].values[0] #.item()
print(most_increase_station, result)


PS 2 306


# 2. 작업형 제2유형
## 멀티클래스 분류 모델

In [2]:
# 필요한 module import
import pandas as pd
import numpy as np
from tabulate import tabulate

# 데이터 불러오기
train = pd.read_csv("/content/yemoonsaBigdata/datasets/Part5/ex6/06.02.01-Metaverse Training Data.csv")
X_test = pd.read_csv("/content/yemoonsaBigdata/datasets/Part5/ex6/06.02.02-Metaverse Testing Data_x.csv")
y_test = pd.read_csv("/content/yemoonsaBigdata/datasets/Part5/ex6/06.02.03-Metaverse Testing Data_y.csv")

# 데이터 탐색
print(train.info())
print(tabulate(train.head(), headers="keys", tablefmt="grid"))
print(X_test.info())
print(tabulate(X_test.head(), headers="keys", tablefmt="grid"))
print(y_test.info())
print(tabulate(y_test.head(), headers="keys", tablefmt="grid"))

# 종속변수 빈도값 확인
print(train['Preferred Metaverse Type'].value_counts())
print(y_test['Preferred Metaverse Type'].value_counts())
# 종속변수값 0,1,2,3 이 비교적 균등하게 분포하고 있다.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Gender                      5000 non-null   object
 1   Education                   5000 non-null   object
 2   Occupation Type             5000 non-null   object
 3   Residence Type              5000 non-null   object
 4   Experience Using Metaverse  5000 non-null   object
 5   Age                         5000 non-null   int64 
 6   Annual Salary               5000 non-null   int64 
 7   Preferred Metaverse Type    5000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 312.6+ KB
None
+----+----------+----------------------------+-------------------------+------------------+------------------------------+-------+-----------------+----------------------------+
|    | Gender   | Education                  | Occupation Type         | Residence Type   | Experie

In [20]:
occupation_counts = train['Occupation Type'].value_counts().reset_index()
occupation_counts.columns = ['Occupation Type', 'Count']

print(tabulate(occupation_counts, headers="keys", tablefmt="grid"))

+----+-------------------------+---------+
|    | Occupation Type         |   Count |
+====+=========================+=========+
|  0 | Fishing industry        |     530 |
+----+-------------------------+---------+
|  1 | Manufacturing industry  |     340 |
+----+-------------------------+---------+
|  2 | Mining                  |     332 |
+----+-------------------------+---------+
|  3 | Service industry        |     318 |
+----+-------------------------+---------+
|  4 | Retail industry         |     317 |
+----+-------------------------+---------+
|  5 | Electric power industry |     314 |
+----+-------------------------+---------+
|  6 | Distribution industry   |     287 |
+----+-------------------------+---------+
|  7 | Wholesale industry      |     270 |
+----+-------------------------+---------+
|  8 | Broadcasting industry   |     267 |
+----+-------------------------+---------+
|  9 | Agriculture             |     264 |
+----+-------------------------+---------+
| 10 | Cons

In [3]:
# 변수 타입별 나누기
COL_DEL =['id'] # 불필요한 변수
COL_CAT = ['Gender', 'Education', 'Occupation Type', 'Residence Type','Experience Using Metaverse'] # 범주형 변수
COL_NUM = ['Age', 'Annual Salary']  # 수치형 변수
COL_Y = ['Preferred Metaverse Type']  # 종속 변수

In [6]:
## train 데이터를 X_train, y_train으로 나누기
X_train = train.drop(train[COL_Y], axis=1)
y_train = train[COL_Y]

In [7]:
## 데이터 분할
from sklearn.model_selection import train_test_split
X_tr, X_val, y_tr, y_val = train_test_split(X_train,
                                            y_train.values.ravel(),
                                            test_size = 0.3,
                                            stratify = y_train.values.ravel())  # 클래스 분포가 훈련세트와 검증세트에서 비슷하게 유지되도록 함

In [11]:
## 수치형변수 - 데이터 스케일링
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X_tr[COL_NUM]) # 데이터 누수를 방지하기 위해 fitting을 학습용 데이터에만 사용하고 해당 결과 다른 데이터에 적용함

X_tr[COL_NUM] = scaler.transform(X_tr[COL_NUM])
X_val[COL_NUM] = scaler.transform(X_val[COL_NUM])
X_test[COL_NUM] = scaler.transform(X_test[COL_NUM])


In [15]:
## 범주형변수 - 데이터 인코딩
from sklearn.preprocessing import LabelEncoder

X = pd.concat([X_train[COL_CAT], X_test[COL_CAT]])

for col in COL_CAT:
  le = LabelEncoder()
  le.fit(X[col])
  X_tr[col] = le.transform(X_tr[col])
  X_val[col] = le.transform(X_val[col])
  X_test[col] = le.transform(X_test[col])

  # 각 변수의 클래스 확인
  print(col)
  print(le.classes_)
  print('\n')

Gender
['female' 'male' 'other']


Education
["Bachelor's degree" 'Doctorate' 'Elementary school graduate'
 'High school graduate' "Master's degree" 'Middle school graduate']


Occupation Type
['Agriculture' 'Broadcasting industry' 'Construction industry'
 'Distribution industry' 'Electric power industry' 'Finance industry'
 'Fishing industry' 'Forestry' 'Gas industry' 'Insurance industry'
 'Maintenance industry' 'Manufacturing industry' 'Mining'
 'Repair industry' 'Retail industry' 'Service industry' 'Water industry'
 'Wholesale industry']


Residence Type
['Apartment' 'Multi-family home' 'Multiple home' 'Officetel'
 'Single-family home' 'Townhouse']


Experience Using Metaverse
['no' 'yes']




In [17]:
print(tabulate(X_tr[COL_CAT].head(), headers="keys", tablefmt='grid'))

+------+----------+-------------+-------------------+------------------+------------------------------+
|      |   Gender |   Education |   Occupation Type |   Residence Type |   Experience Using Metaverse |
+======+==========+=============+===================+==================+==============================+
| 2778 |        2 |           0 |                10 |                0 |                            0 |
+------+----------+-------------+-------------------+------------------+------------------------------+
| 1629 |        1 |           5 |                15 |                4 |                            0 |
+------+----------+-------------+-------------------+------------------+------------------------------+
| 4230 |        1 |           1 |                 6 |                0 |                            0 |
+------+----------+-------------+-------------------+------------------+------------------------------+
| 4703 |        2 |           3 |                14 |           

In [21]:
## 모델 학습
## 1) 로지스틱 리그레션
from sklearn.linear_model import LogisticRegression
modelLR = LogisticRegression(random_state = 123)
modelLR.fit(X_tr, y_tr)

## 2) 랜덤 포레스트
from sklearn.ensemble import RandomForestClassifier
modelRF = RandomForestClassifier(random_state = 123)
modelRF.fit(X_tr, y_tr)

## 3) XGBoost
from xgboost import XGBClassifier
modelXGB = XGBClassifier(random_state = 123)
modelXGB.fit(X_tr, y_tr)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, objective='multi:softprob', ...)

In [23]:
## 모델 평가
y_val_predLR = modelLR.predict(X_val)
y_val_predRF = modelRF.predict(X_val)
y_val_predXGB = modelXGB.predict(X_val)

## F1 Score 사용
from sklearn.metrics import f1_score, accuracy_score

scoreLR = f1_score(y_val, y_val_predLR, average='macro')
scoreRF = f1_score(y_val, y_val_predRF, average='macro')
scoreXGB = f1_score(y_val, y_val_predXGB, average='macro')

print('Logistic Regression: \t', scoreLR)
print('Random Forest: \t', scoreRF)
print('XGBoost: \t', scoreXGB)

Logistic Regression: 	 0.27049185869757564
Random Forest: 	 0.9685734201098353
XGBoost: 	 0.9685618525081114


In [24]:
## 참고 - metric 종류 확인
import sklearn.metrics
print(help(sklearn.metrics.f1_score))

Help on function f1_score in module sklearn.metrics._classification:

f1_score(y_true, y_pred, *, labels=None, pos_label=1, average='binary', sample_weight=None, zero_division='warn')
    Compute the F1 score, also known as balanced F-score or F-measure.
    
    The F1 score can be interpreted as a harmonic mean of the precision and
    recall, where an F1 score reaches its best value at 1 and worst score at 0.
    The relative contribution of precision and recall to the F1 score are
    equal. The formula for the F1 score is::
    
        F1 = 2 * (precision * recall) / (precision + recall)
    
    In the multi-class and multi-label case, this is the average of
    the F1 score of each class with weighting depending on the ``average``
    parameter.
    
    Read more in the :ref:`User Guide <precision_recall_f_measure_metrics>`.
    
    Parameters
    ----------
    y_true : 1d array-like, or label indicator array / sparse matrix
        Ground truth (correct) target values.
    

In [27]:
## 학습, 검증 데이터 성능 확인 함수
def get_scores(model, X_tr, X_val, y_tr, y_val):
  y_tr_pred = model.predict(X_tr)
  y_val_pred = model.predict(X_val)
  tr_f1_score = f1_score(y_tr, y_tr_pred, average='macro')
  val_f1_score = f1_score(y_val, y_val_pred, average='macro')
  tr_acc_score = accuracy_score(y_tr, y_tr_pred)
  val_acc_score = accuracy_score(y_val, y_val_pred)

  return f'F1_train: {round(tr_f1_score, 4)}, Acc_train: {round(tr_acc_score, 4)}, F1_valid: {round(val_f1_score, 4)}, Acc_valid: {round(val_acc_score, 4)} '


In [28]:
print('Logistic Regression: \t', get_scores(modelLR, X_tr, X_val, y_tr, y_val))
print('Random Forest: \t', get_scores(modelRF, X_tr, X_val, y_tr, y_val))
print('XGBoost: \t', get_scores(modelXGB, X_tr, X_val, y_tr, y_val))

Logistic Regression: 	 F1_train: 0.2873, Acc_train: 0.2966, F1_valid: 0.2705, Acc_valid: 0.2807 
Random Forest: 	 F1_train: 1.0, Acc_train: 1.0, F1_valid: 0.9686, Acc_valid: 0.9687 
XGBoost: 	 F1_train: 1.0, Acc_train: 1.0, F1_valid: 0.9686, Acc_valid: 0.9687 


In [29]:
## 하이퍼파라미터 튜닝
modelXGB2 = XGBClassifier(n_estimators=50, max_depth=3, min_child_weight=1, random_state=123)
modelXGB2.fit(X_tr, y_tr)

modelXGB3 = XGBClassifier(n_estimators=50, max_depth=3, min_child_weight=2, random_state=123)
modelXGB3.fit(X_tr, y_tr)

modelXGB4 = XGBClassifier(n_estimators=50, max_depth=5, min_child_weight=1, random_state=123)
modelXGB4.fit(X_tr, y_tr)

modelXGB5 = XGBClassifier(n_estimators=100, max_depth=5, min_child_weight=1, random_state=123)
modelXGB5.fit(X_tr, y_tr)

print('XGBoost2: \t', get_scores(modelXGB2, X_tr, X_val, y_tr, y_val))
print('XGBoost3: \t', get_scores(modelXGB3, X_tr, X_val, y_tr, y_val))
print('XGBoost4: \t', get_scores(modelXGB4, X_tr, X_val, y_tr, y_val))
print('XGBoost5: \t', get_scores(modelXGB5, X_tr, X_val, y_tr, y_val))

XGBoost2: 	 F1_train: 0.7922, Acc_train: 0.7931, F1_valid: 0.6908, Acc_valid: 0.6933 
XGBoost3: 	 F1_train: 0.7811, Acc_train: 0.7817, F1_valid: 0.6752, Acc_valid: 0.6773 
XGBoost4: 	 F1_train: 0.9703, Acc_train: 0.9703, F1_valid: 0.9138, Acc_valid: 0.914 
XGBoost5: 	 F1_train: 0.9963, Acc_train: 0.9963, F1_valid: 0.9627, Acc_valid: 0.9627 


In [32]:
## 최종 결과값 생성
pred = modelXGB5.predict(X_test[COL_CAT+COL_NUM])
result = pd.DataFrame({'ID': X_test.id, 'Segmentation': pred})
print(result.head())

result.to_csv('hansjang.csv', index=False)

   ID  Segmentation
0   1             1
1   2             1
2   3             1
3   4             0
4   5             0


# 3. 작업형 제3유형

## 01. 교통수단 : 카이제곱 적합도 검정
### https://m.blog.naver.com/gksshdk8003/222305638201

In [33]:
import pandas as pd
import numpy as np
from tabulate import tabulate
import scipy.stats as stats

df = pd.read_csv('/content/yemoonsaBigdata/datasets/Part5/ex6/06.03.01-transport_data.csv')

print(tabulate(df.head(), headers="keys", tablefmt="grid"))

+----+-----------+-----------------+
|    | User ID   | TransportMode   |
+====+===========+=================+
|  0 | U1        | subway          |
+----+-----------+-----------------+
|  1 | U2        | taxi            |
+----+-----------+-----------------+
|  2 | U3        | subway          |
+----+-----------+-----------------+
|  3 | U4        | subway          |
+----+-----------+-----------------+
|  4 | U5        | subway          |
+----+-----------+-----------------+


In [50]:
observed_data = df['TransportMode'].value_counts().reset_index()
observed_data.columns = ['TransportMode', 'Count']

print(tabulate(observed_data, headers="keys", tablefmt="grid"))

+----+-----------------+---------+
|    | TransportMode   |   Count |
+====+=================+=========+
|  0 | bus             |     415 |
+----+-----------------+---------+
|  1 | subway          |     294 |
+----+-----------------+---------+
|  2 | taxi            |     183 |
+----+-----------------+---------+
|  3 | bicycle         |     108 |
+----+-----------------+---------+


In [51]:
## 지하철 선호 이용자 비율 구하기
total_counts = sum(observed_data['Count'])
subway_counts = observed_data[observed_data['TransportMode']=='subway']['Count'].values[0]
subway_ratio = subway_counts / total_counts
print(round(subway_ratio, 2))

0.29


In [52]:
expected_data = {
    'TransportMode' : ['bus', 'subway', 'taxi', 'bicycle'],
    'Count' : [total_counts * 0.4, total_counts * 0.3, total_counts * 0.2, total_counts * 0.1]
}

expected_data = pd.DataFrame(expected_data)

print(tabulate(expected_data, headers="keys", tablefmt="grid"))

+----+-----------------+---------+
|    | TransportMode   |   Count |
+====+=================+=========+
|  0 | bus             |     400 |
+----+-----------------+---------+
|  1 | subway          |     300 |
+----+-----------------+---------+
|  2 | taxi            |     200 |
+----+-----------------+---------+
|  3 | bicycle         |     100 |
+----+-----------------+---------+


In [53]:
# 관찰빈도 만들기
observed_counts = observed_data['Count'].values
print(observed_counts)

# 기대빈도 만들기
expected_counts = expected_data['Count'].values
print(expected_counts)

[415 294 183 108]
[400. 300. 200. 100.]


In [55]:
# 검정통계량, p-value 구하기
stat, pvalue = stats.chisquare(list(observed_counts), list(expected_counts))
print(round(stat, 2), round(pvalue, 2))

2.77 0.43


p-value가 0.05보다 크기 때문에 귀무가설을 채택한다. 즉, 대중교통 이용객의 실제 교통수단 선택 비율이 에상 비율과 같다.

## 02. 물고기 무게 : 다중회귀분석

In [58]:
import pandas as pd
import statsmodels.api as sm
from tabulate import tabulate as tb

df = pd.read_csv('/content/yemoonsaBigdata/datasets/Part5/ex6/06.03.02-fish_weight_data.csv')

print(tb(df.head(), headers="keys", tablefmt="grid"))


+----+------+----------+----------+---------+-----------+----------+
|    |   id |   Length |   Height |   Width |   FinSize |   Weight |
+====+======+==========+==========+=========+===========+==========+
|  0 |    1 |    63.91 |    28.72 |    3.46 |     18.22 |   194.32 |
+----+------+----------+----------+---------+-----------+----------+
|  1 |    2 |    77.22 |    14.45 |    7.12 |     15.71 |   194.59 |
+----+------+----------+----------+---------+-----------+----------+
|  2 |    3 |    68.22 |    30.73 |    4.09 |      7.33 |   191.24 |
+----+------+----------+----------+---------+-----------+----------+
|  3 |    4 |    63.59 |    38.68 |    2.21 |      2.54 |   194.23 |
+----+------+----------+----------+---------+-----------+----------+
|  4 |    5 |    53.89 |    13.71 |    0.73 |      8.74 |   131.82 |
+----+------+----------+----------+---------+-----------+----------+


In [59]:
## 독립변수와 종속변수 설정
X_cols = ['Length', 'Height', 'Width', 'FinSize']
y_cols = ['Weight']

## 독립변수와 종속변수 분리
X = df[X_cols]
y = df[y_cols]

print(tb(X.head(), headers='keys', tablefmt='grid'))

+----+----------+----------+---------+-----------+
|    |   Length |   Height |   Width |   FinSize |
+====+==========+==========+=========+===========+
|  0 |    63.91 |    28.72 |    3.46 |     18.22 |
+----+----------+----------+---------+-----------+
|  1 |    77.22 |    14.45 |    7.12 |     15.71 |
+----+----------+----------+---------+-----------+
|  2 |    68.22 |    30.73 |    4.09 |      7.33 |
+----+----------+----------+---------+-----------+
|  3 |    63.59 |    38.68 |    2.21 |      2.54 |
+----+----------+----------+---------+-----------+
|  4 |    53.89 |    13.71 |    0.73 |      8.74 |
+----+----------+----------+---------+-----------+


In [66]:
## 다중회귀 분석 수행
X = sm.add_constant(X)  # 상수항 추가
model = sm.OLS(y, X).fit()  # 분석 수행 - 모델 생성

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 Weight   R-squared:                       0.991
Model:                            OLS   Adj. R-squared:                  0.991
Method:                 Least Squares   F-statistic:                     2751.
Date:                Tue, 27 Aug 2024   Prob (F-statistic):           2.96e-97
Time:                        09:50:51   Log-Likelihood:                -292.18
No. Observations:                 100   AIC:                             594.4
Df Residuals:                      95   BIC:                             607.4
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -3.3186      2.278     -1.457      0.1

In [61]:
## Length 회귀계수 구하기 length_coef
length_coef = model.params["Length"]
length_coef = round(length_coef, 2)

print(length_coef)

2.04


In [65]:
## Height 계수의 p-value 구하기

height_pvalue = model.pvalues['Height']
print(height_pvalue)

2.0296276104742017e-51


In [67]:
## Width 회귀계수의 유의미 여부 추출하기
width_coef = model.params["Width"]
width_pvalue = model.pvalues['Width']

if width_pvalue < 0.05:
  print('Y')
else:
  print('N')

Y
